src/sparseconverter/__init__.py

```diff
@@ -377,67 +376,68 @@ def _build_converters(self):
                 for right in NUMPY, CUDA:
-                    if (left, right) not in self._converters:
-                        self._converters[(left, right)] = _classes[left].toarray
-            for left in SCIPY_CSR, SCIPY_CSC:
+                    self._complete((left, right), _classes[left].toarray)
+            for left in SCIPY_CSR, SCIPY_CSC, SCIPY_CSR_ARRAY, SCIPY_CSC_ARRAY:
                 for right in SPARSE_COO, SPARSE_GCXS, SPARSE_DOK:
-                    if (left, right) not in self._converters:
-                        self._converters[(left, right)] = _convert_csc_csr_to_pydata_sparse(
-                            left, right
-                        )
+                    self._complete(
+                        (left, right),
+                        _convert_csc_csr_to_pydata_sparse(left, right)
+                    )
             for left in SCIPY_COO, :
                 for right in SPARSE_COO, SPARSE_GCXS, SPARSE_DOK:
-                    if (left, right) not in self._converters:
-                        self._converters[(left, right)] = _classes[right].from_scipy_sparse
-            self._converters[(SPARSE_COO, SCIPY_COO)] = chain(_flatsig, sparse.COO.to_scipy_sparse)
-            self._converters[(SPARSE_COO, SCIPY_CSR)] = chain(_flatsig, sparse.COO.tocsr)
-            self._converters[(SPARSE_COO, SCIPY_CSC)] = chain(_flatsig, sparse.COO.tocsc)
-            self._converters[(SPARSE_GCXS, SCIPY_COO)] = _GCXS_to_coo
-            self._converters[(SPARSE_GCXS, SCIPY_CSR)] = _GCXS_to_csr
-            self._converters[(SPARSE_GCXS, SCIPY_CSC)] = _GCXS_to_csc
+                    self._complete((left, right), _classes[right].from_scipy_sparse)
+
+            self._insert((SPARSE_COO, SCIPY_COO), chain(_flatsig, sparse.COO.to_scipy_sparse))
+            self._insert((SPARSE_COO, SCIPY_CSR), chain(_flatsig, sparse.COO.tocsr))
+            self._insert((SPARSE_COO, SCIPY_CSC), chain(_flatsig, sparse.COO.tocsc))
+            self._insert((SPARSE_GCXS, SCIPY_COO), _GCXS_to_coo)
+            self._insert((SPARSE_GCXS, SCIPY_CSR), _GCXS_to_csr)
+            self._insert((SPARSE_GCXS, SCIPY_CSC), _GCXS_to_csc)
 
             for right in SCIPY_COO, SCIPY_CSR, SCIPY_CSC:
-                if (SPARSE_DOK, right) not in self._converters:
-                    self._converters[(SPARSE_DOK, right)] = chain(
+                self._complete(
+                    (SPARSE_DOK, right),
+                    chain(
                         self._converters[(SPARSE_DOK, SPARSE_COO)],
                         self._converters[(SPARSE_COO, right)]
                     )
-            # Make sure scipy.sparse.*_array entries are populated before
+                )
+            # Make sure missing scipy.sparse.*_array entries are populated before
             # following stage
             self._populate_scipy_array()
 
-            for left in CPU_BACKENDS:
+            for left in CPU_BACKENDS.union((CUDA, )):
```

Comment:
Can you explain, maybe in a comment, why we have `CUDA` in here? At first look, this looks like a "all backends" loop, but that's not the case, is it? It's excluding the cupy formats, correct? Why is this the case?

Reply:
I've introduced a new top-level set of backends without CuPy that should make this clearer.